From fb090575d08760f1cc88f6492ce973bb118fec08 Mon Sep 17 00:00:00 2001
From: "Zhibin Wu (rebased)" <zbwu@rebase>
Date: Sun, 19 Jul 2026 11:32:28 +0800
Subject: [PATCH 1/4] Athena DTS: initial support + regulator/camera update,
 restructured for r35.6.4 kernel-dts layout

Squashes zbwu's two athena_l4t_jakku_dts commits (389f95e, b994542) and moves
the files from the r35.1 repo-root layout into r35.6.4 kernel-dts/. Stock
p3509 devkit DTBs stay enabled (zbwu's BUILD_ATHENA gating dropped).
---
 kernel-dts/Makefile                           |   1 +
 kernel-dts/common/tegra194-camera-p2151.dtsi  | 790 ++++++++++++++++++
 .../tegra194-fixed-regulator-p2151-0000.dtsi  | 225 +++++
 kernel-dts/common/tegra194-p2151-0000.dtsi    | 456 ++++++++++
 kernel-dts/common/tegra194-p2151-disp.dtsi    |  61 ++
 kernel-dts/tegra194-p3668-0001-p2151-0000.dts |  26 +
 6 files changed, 1559 insertions(+)
 create mode 100644 kernel-dts/common/tegra194-camera-p2151.dtsi
 create mode 100644 kernel-dts/common/tegra194-fixed-regulator-p2151-0000.dtsi
 create mode 100644 kernel-dts/common/tegra194-p2151-0000.dtsi
 create mode 100644 kernel-dts/common/tegra194-p2151-disp.dtsi
 create mode 100644 kernel-dts/tegra194-p3668-0001-p2151-0000.dts

diff --git a/kernel-dts/Makefile b/kernel-dts/Makefile
index 925e46f..0fff100 100644
--- a/kernel-dts/Makefile
+++ b/kernel-dts/Makefile
@@ -12,6 +12,7 @@ endif
 dtb-$(BUILD_ENABLE) += tegra194-p3668-all-p3509-0000.dtb
 dtb-$(BUILD_ENABLE) += tegra194-p3668-0000-p3509-0000.dtb
 dtb-$(BUILD_ENABLE) += tegra194-p3668-0001-p3509-0000.dtb
+dtb-$(BUILD_ENABLE) += tegra194-p3668-0001-p2151-0000.dtb
 dtb-$(CONFIG_ARCH_TEGRA_19x_SOC) += tegra194-p3668-all-p3509-0000-kexec.dtb
 dtbo-$(BUILD_ENABLE) += tegra194-p3668-all-p3509-0000-hdr40.dtbo
 dtbo-$(BUILD_ENABLE) += tegra194-p3668-all-p3509-0000-adafruit-sph0645lm4h.dtbo
diff --git a/kernel-dts/common/tegra194-camera-p2151.dtsi b/kernel-dts/common/tegra194-camera-p2151.dtsi
new file mode 100644
index 0000000..4fc0baa
--- /dev/null
+++ b/kernel-dts/common/tegra194-camera-p2151.dtsi
@@ -0,0 +1,790 @@
+/*
+ * Copyright (c) 2015-2020, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms and conditions of the GNU General Public License,
+ * version 2, as published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/ {
+	tegra-capture-vi {
+		num-channels = <3>;
+		ports {
+			#address-cells = <1>;
+			#size-cells = <0>;
+			port@0 {
+				reg = <0>;
+				p2151_vi_in0: endpoint {
+					port-index = <0>;
+					bus-width = <1>;
+					remote-endpoint = <&p2151_csi_out0>;
+				};
+			};
+			port@1 {
+				reg = <1>;
+				p2151_vi_in1: endpoint {
+					port-index = <1>;
+					bus-width = <1>;
+					remote-endpoint = <&p2151_csi_out1>;
+				};
+			};
+			port@2 {
+				reg = <2>;
+				p2151_vi_in2: endpoint {
+					port-index = <4>;
+					bus-width = <4>;
+					remote-endpoint = <&p2151_csi_out2>;
+				};
+			};
+		};
+	};
+
+	host1x@13e00000 {
+		nvcsi@15a00000 {
+			num-channels = <3>;
+			#address-cells = <1>;
+			#size-cells = <0>;
+			channel@0 {
+				reg = <0>;
+				ports {
+					#address-cells = <1>;
+					#size-cells = <0>;
+					port@0 {
+						reg = <0>;
+						p2151_csi_in0: endpoint@0 {
+							port-index = <0>;
+							bus-width = <1>;
+							remote-endpoint = <&p2151_ov7251_out0>;
+						};
+					};
+					port@1 {
+						reg = <1>;
+						p2151_csi_out0: endpoint@1 {
+							remote-endpoint = <&p2151_vi_in0>;
+						};
+					};
+				};
+			};
+			channel@1 {
+				reg = <1>;
+				ports {
+					#address-cells = <1>;
+					#size-cells = <0>;
+					port@0 {
+						reg = <0>;
+						p2151_csi_in1: endpoint@2 {
+							port-index = <1>;
+							bus-width = <1>;
+							remote-endpoint = <&p2151_ov7251_out1>;
+						};
+					};
+					port@1 {
+						reg = <1>;
+						p2151_csi_out1: endpoint@3 {
+							remote-endpoint = <&p2151_vi_in1>;
+						};
+					};
+				};
+			};
+			channel@2 {
+				reg = <2>;
+				ports {
+					#address-cells = <1>;
+					#size-cells = <0>;
+					port@0 {
+						reg = <0>;
+						p2151_csi_in2: endpoint@4 {
+							port-index = <4>;
+							bus-width = <4>;
+							remote-endpoint = <&p2151_ov13b10_out2>;
+						};
+					};
+					port@1 {
+						reg = <1>;
+						p2151_csi_out2: endpoint@5 {
+							remote-endpoint = <&p2151_vi_in2>;
+						};
+					};
+				};
+			};
+		};
+	};
+
+	i2c@3180000 {
+		ov7251_a@61 {
+			compatible = "nvidia,ov7251";
+			/* I2C device address */
+			reg = <0x61>;
+
+			/* Physical dimensions of sensor */
+			physical_w = "1.968";
+			physical_h = "1.488";
+
+			/* Define any required hw resources needed by driver */
+			/* ie. clocks, io pins, power sources */
+			avdd-reg = "vana";
+			iovdd-reg = "vif";
+
+			sensor_model = "ov7251";
+			use_sensor_mode_id = "true";
+
+			/* Sensor output flip settings */
+			vertical-flip = "true";
+			horizontal-mirror = "true";
+
+			/**
+			* A modeX node is required to support v4l2 driver
+			* implementation with NVIDIA camera software stack
+			*
+			* mclk_khz = "";
+			* Standard MIPI driving clock, typically 24MHz
+			*
+			* num_lanes = "";
+			* Number of lane channels sensor is programmed to output
+			*
+			* tegra_sinterface = "";
+			* The base tegra serial interface lanes are connected to
+			* Incase of virtual HW devices, use virtual
+			* For SW emulated devices, use host
+			*
+			* phy_mode = "";
+			* PHY mode used by the MIPI lanes for this device
+			*
+			* discontinuous_clk = "";
+			* The sensor is programmed to use a discontinuous clock on MIPI lanes
+			*
+			* dpcm_enable = "true";
+			* The sensor is programmed to use a DPCM modes
+			*
+			* cil_settletime = "";
+			* MIPI lane settle time value.
+			* A "0" value attempts to autocalibrate based on mclk_multiplier
+			*
+			*
+			*
+			*
+			* active_w = "";
+			* Pixel active region width
+			*
+			* active_h = "";
+			* Pixel active region height
+			*
+			* pixel_t = "";
+			* The sensor readout pixel pattern
+			*
+			* readout_orientation = "0";
+			* Based on camera module orientation.
+			* Only change readout_orientation if you specifically
+			* Program a different readout order for this mode
+			*
+			* line_length = "";
+			* Pixel line length (width) for sensor mode.
+			* This is used to calibrate features in our camera stack.
+			*
+			* mclk_multiplier = "";
+			* Multiplier to MCLK to help time hardware capture sequence
+			* TODO: Assign to PLL_Multiplier as well until fixed in core
+			*
+			* pix_clk_hz = "";
+			* Sensor pixel clock used for calculations like exposure and framerate
+			*
+			*
+			*
+			*
+			* inherent_gain = "";
+			* Gain obtained inherently from mode (ie. pixel binning)
+			*
+			* == Source Control Settings ==
+			*
+			* Gain factor used to convert fixed point integer to float
+			* Gain range [min_gain/gain_factor, max_gain/gain_factor]
+			* Gain step [step_gain/gain_factor is the smallest step that can be configured]
+			* Default gain [Default gain to be initialized for the control.
+			*     use min_gain_val as default for optimal results]
+			* Framerate factor used to convert fixed point integer to float
+			* Framerate range [min_framerate/framerate_factor, max_framerate/framerate_factor]
+			* Framerate step [step_framerate/framerate_factor is the smallest step that can be configured]
+			* Default Framerate [Default framerate to be initialized for the control.
+			*     use max_framerate to get required performance]
+			* Exposure factor used to convert fixed point integer to float
+			* For convenience use 1 sec = 1000000us as conversion factor
+			* Exposure range [min_exp_time/exposure_factor, max_exp_time/exposure_factor]
+			* Exposure step [step_exp_time/exposure_factor is the smallest step that can be configured]
+			* Default Exposure Time [Default exposure to be initialized for the control.
+			*     Set default exposure based on the default_framerate for optimal exposure settings]
+			*
+			* gain_factor = ""; (integer factor used for floating to fixed point conversion)
+			* min_gain_val = ""; (ceil to integer)
+			* max_gain_val = ""; (ceil to integer)
+			* step_gain_val = ""; (ceil to integer)
+			* default_gain = ""; (ceil to integer)
+			* Gain limits for mode
+			*
+			* exposure_factor = ""; (integer factor used for floating to fixed point conversion)
+			* min_exp_time = ""; (ceil to integer)
+			* max_exp_time = ""; (ceil to integer)
+			* step_exp_time = ""; (ceil to integer)
+			* default_exp_time = ""; (ceil to integer)
+			* Exposure Time limits for mode (us)
+			*
+			*
+			* min_hdr_ratio = "";
+			* max_hdr_ratio = "";
+			* HDR Ratio limits for mode
+			*
+			* framerate_factor = ""; (integer factor used for floating to fixed point conversion)
+			* min_framerate = "";
+			* max_framerate = "";
+			* step_framerate = ""; (ceil to integer)
+			* default_framerate = ""; (ceil to integer)
+			* Framerate limits for mode (fps)
+			*/
+			mode0 { // OV7251_MODE_640X480
+				mclk_khz = "24000";
+				num_lanes = "1";
+				tegra_sinterface = "serial_a";
+				phy_mode = "DPHY";
+				discontinuous_clk = "yes";
+				dpcm_enable = "false";
+				cil_settletime = "0";
+
+				active_w = "640";
+				active_h = "480";
+				mode_type = "bayer";
+				pixel_phase = "bggr";
+				csi_pixel_bit_depth = "10";
+				readout_orientation = "0";
+				line_length = "928";
+				inherent_gain = "1";
+				mclk_multiplier = "3.33";
+				pix_clk_hz = "80000000";
+
+				gain_factor = "16";
+				min_gain_val = "16"; /* 1.00x */
+				max_gain_val = "248"; /* 15.5x */
+				step_gain_val = "1";
+				default_gain = "16"; /* 1.00x */
+				min_hdr_ratio = "1";
+				max_hdr_ratio = "1";
+				framerate_factor = "1000000";
+				min_framerate = "30000000"; /* 30.0 fps */
+				max_framerate = "30000000"; /* 30.0 fps */
+				step_framerate = "1";
+				default_framerate = "30000000"; /* 30.0 fps */
+				exposure_factor = "1000000";
+				min_exp_time = "19"; /* us */
+				max_exp_time = "32943"; /* us */
+				step_exp_time = "1";
+				default_exp_time = "32943";/* us */
+				embedded_metadata_height = "0";
+			};
+
+			ports {
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				port@0 {
+					reg = <0>;
+					p2151_ov7251_out0: endpoint {
+						port-index = <0>;
+						bus-width = <1>;
+						remote-endpoint = <&p2151_csi_in0>;
+					};
+				};
+			};
+		};
+		ov7251_b@62 {
+			compatible = "nvidia,ov7251";
+			/* I2C device address */
+			reg = <0x62>;
+
+			/* Physical dimensions of sensor */
+			physical_w = "1.968";
+			physical_h = "1.488";
+
+			/* Define any required hw resources needed by driver */
+			/* ie. clocks, io pins, power sources */
+			avdd-reg = "vana";
+			iovdd-reg = "vif";
+
+			sensor_model = "ov7251";
+			use_sensor_mode_id = "true";
+
+			/* Sensor output flip settings */
+			vertical-flip = "true";
+			horizontal-mirror = "true";
+
+			/**
+			* A modeX node is required to support v4l2 driver
+			* implementation with NVIDIA camera software stack
+			*
+			* mclk_khz = "";
+			* Standard MIPI driving clock, typically 24MHz
+			*
+			* num_lanes = "";
+			* Number of lane channels sensor is programmed to output
+			*
+			* tegra_sinterface = "";
+			* The base tegra serial interface lanes are connected to
+			* Incase of virtual HW devices, use virtual
+			* For SW emulated devices, use host
+			*
+			* phy_mode = "";
+			* PHY mode used by the MIPI lanes for this device
+			*
+			* discontinuous_clk = "";
+			* The sensor is programmed to use a discontinuous clock on MIPI lanes
+			*
+			* dpcm_enable = "true";
+			* The sensor is programmed to use a DPCM modes
+			*
+			* cil_settletime = "";
+			* MIPI lane settle time value.
+			* A "0" value attempts to autocalibrate based on mclk_multiplier
+			*
+			*
+			*
+			*
+			* active_w = "";
+			* Pixel active region width
+			*
+			* active_h = "";
+			* Pixel active region height
+			*
+			* pixel_t = "";
+			* The sensor readout pixel pattern
+			*
+			* readout_orientation = "0";
+			* Based on camera module orientation.
+			* Only change readout_orientation if you specifically
+			* Program a different readout order for this mode
+			*
+			* line_length = "";
+			* Pixel line length (width) for sensor mode.
+			* This is used to calibrate features in our camera stack.
+			*
+			* mclk_multiplier = "";
+			* Multiplier to MCLK to help time hardware capture sequence
+			* TODO: Assign to PLL_Multiplier as well until fixed in core
+			*
+			* pix_clk_hz = "";
+			* Sensor pixel clock used for calculations like exposure and framerate
+			*
+			*
+			*
+			*
+			* inherent_gain = "";
+			* Gain obtained inherently from mode (ie. pixel binning)
+			*
+			* == Source Control Settings ==
+			*
+			* Gain factor used to convert fixed point integer to float
+			* Gain range [min_gain/gain_factor, max_gain/gain_factor]
+			* Gain step [step_gain/gain_factor is the smallest step that can be configured]
+			* Default gain [Default gain to be initialized for the control.
+			*     use min_gain_val as default for optimal results]
+			* Framerate factor used to convert fixed point integer to float
+			* Framerate range [min_framerate/framerate_factor, max_framerate/framerate_factor]
+			* Framerate step [step_framerate/framerate_factor is the smallest step that can be configured]
+			* Default Framerate [Default framerate to be initialized for the control.
+			*     use max_framerate to get required performance]
+			* Exposure factor used to convert fixed point integer to float
+			* For convenience use 1 sec = 1000000us as conversion factor
+			* Exposure range [min_exp_time/exposure_factor, max_exp_time/exposure_factor]
+			* Exposure step [step_exp_time/exposure_factor is the smallest step that can be configured]
+			* Default Exposure Time [Default exposure to be initialized for the control.
+			*     Set default exposure based on the default_framerate for optimal exposure settings]
+			*
+			* gain_factor = ""; (integer factor used for floating to fixed point conversion)
+			* min_gain_val = ""; (ceil to integer)
+			* max_gain_val = ""; (ceil to integer)
+			* step_gain_val = ""; (ceil to integer)
+			* default_gain = ""; (ceil to integer)
+			* Gain limits for mode
+			*
+			* exposure_factor = ""; (integer factor used for floating to fixed point conversion)
+			* min_exp_time = ""; (ceil to integer)
+			* max_exp_time = ""; (ceil to integer)
+			* step_exp_time = ""; (ceil to integer)
+			* default_exp_time = ""; (ceil to integer)
+			* Exposure Time limits for mode (us)
+			*
+			*
+			* min_hdr_ratio = "";
+			* max_hdr_ratio = "";
+			* HDR Ratio limits for mode
+			*
+			* framerate_factor = ""; (integer factor used for floating to fixed point conversion)
+			* min_framerate = "";
+			* max_framerate = "";
+			* step_framerate = ""; (ceil to integer)
+			* default_framerate = ""; (ceil to integer)
+			* Framerate limits for mode (fps)
+			*/
+			mode0 { // OV7251_MODE_640X480
+				mclk_khz = "24000";
+				num_lanes = "1";
+				tegra_sinterface = "serial_b";
+				phy_mode = "DPHY";
+				discontinuous_clk = "yes";
+				dpcm_enable = "false";
+				cil_settletime = "0";
+
+				active_w = "640";
+				active_h = "480";
+				mode_type = "bayer";
+				pixel_phase = "bggr";
+				csi_pixel_bit_depth = "10";
+				readout_orientation = "0";
+				line_length = "928";
+				inherent_gain = "1";
+				mclk_multiplier = "3.33";
+				pix_clk_hz = "80000000";
+
+				gain_factor = "16";
+				min_gain_val = "16"; /* 1.00x */
+				max_gain_val = "248"; /* 15.5x */
+				step_gain_val = "1";
+				default_gain = "16"; /* 1.00x */
+				min_hdr_ratio = "1";
+				max_hdr_ratio = "1";
+				framerate_factor = "1000000";
+				min_framerate = "30000000"; /* 30.0 fps */
+				max_framerate = "30000000"; /* 30.0 fps */
+				step_framerate = "1";
+				default_framerate = "30000000"; /* 30.0 fps */
+				exposure_factor = "1000000";
+				min_exp_time = "19"; /* us */
+				max_exp_time = "32943"; /* us */
+				step_exp_time = "1";
+				default_exp_time = "32943";/* us */
+				embedded_metadata_height = "0";
+			};
+
+			ports {
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				port@0 {
+					reg = <0>;
+					p2151_ov7251_out1: endpoint {
+						port-index = <1>;
+						bus-width = <1>;
+						remote-endpoint = <&p2151_csi_in1>;
+					};
+				};
+			};
+		};
+
+		ov13b10_e@36 {
+			compatible = "nvidia,ov13b10";
+			/* I2C device address */
+			reg = <0x36>;
+
+			/* Physical dimensions of sensor */
+			physical_w = "4.713984";
+			physical_h = "3.499776";
+
+			/* Define any required hw resources needed by driver */
+			/* ie. clocks, io pins, power sources */
+			avdd-reg = "vana";
+			iovdd-reg = "vif";
+
+			sensor_model = "ov13b10";
+			use_sensor_mode_id = "true";
+
+			/* Sensor output flip settings */
+			vertical-flip = "true";
+			horizontal-mirror = "true";
+
+			/**
+			* A modeX node is required to support v4l2 driver
+			* implementation with NVIDIA camera software stack
+			*
+			* mclk_khz = "";
+			* Standard MIPI driving clock, typically 24MHz
+			*
+			* num_lanes = "";
+			* Number of lane channels sensor is programmed to output
+			*
+			* tegra_sinterface = "";
+			* The base tegra serial interface lanes are connected to
+			* Incase of virtual HW devices, use virtual
+			* For SW emulated devices, use host
+			*
+			* phy_mode = "";
+			* PHY mode used by the MIPI lanes for this device
+			*
+			* discontinuous_clk = "";
+			* The sensor is programmed to use a discontinuous clock on MIPI lanes
+			*
+			* dpcm_enable = "true";
+			* The sensor is programmed to use a DPCM modes
+			*
+			* cil_settletime = "";
+			* MIPI lane settle time value.
+			* A "0" value attempts to autocalibrate based on mclk_multiplier
+			*
+			*
+			*
+			*
+			* active_w = "";
+			* Pixel active region width
+			*
+			* active_h = "";
+			* Pixel active region height
+			*
+			* pixel_t = "";
+			* The sensor readout pixel pattern
+			*
+			* readout_orientation = "0";
+			* Based on camera module orientation.
+			* Only change readout_orientation if you specifically
+			* Program a different readout order for this mode
+			*
+			* line_length = "";
+			* Pixel line length (width) for sensor mode.
+			* This is used to calibrate features in our camera stack.
+			*
+			* mclk_multiplier = "";
+			* Multiplier to MCLK to help time hardware capture sequence
+			* TODO: Assign to PLL_Multiplier as well until fixed in core
+			*
+			* pix_clk_hz = "";
+			* Sensor pixel clock used for calculations like exposure and framerate
+			*
+			*
+			*
+			*
+			* inherent_gain = "";
+			* Gain obtained inherently from mode (ie. pixel binning)
+			*
+			* == Source Control Settings ==
+			*
+			* Gain factor used to convert fixed point integer to float
+			* Gain range [min_gain/gain_factor, max_gain/gain_factor]
+			* Gain step [step_gain/gain_factor is the smallest step that can be configured]
+			* Default gain [Default gain to be initialized for the control.
+			*     use min_gain_val as default for optimal results]
+			* Framerate factor used to convert fixed point integer to float
+			* Framerate range [min_framerate/framerate_factor, max_framerate/framerate_factor]
+			* Framerate step [step_framerate/framerate_factor is the smallest step that can be configured]
+			* Default Framerate [Default framerate to be initialized for the control.
+			*     use max_framerate to get required performance]
+			* Exposure factor used to convert fixed point integer to float
+			* For convenience use 1 sec = 1000000us as conversion factor
+			* Exposure range [min_exp_time/exposure_factor, max_exp_time/exposure_factor]
+			* Exposure step [step_exp_time/exposure_factor is the smallest step that can be configured]
+			* Default Exposure Time [Default exposure to be initialized for the control.
+			*     Set default exposure based on the default_framerate for optimal exposure settings]
+			*
+			* gain_factor = ""; (integer factor used for floating to fixed point conversion)
+			* min_gain_val = ""; (ceil to integer)
+			* max_gain_val = ""; (ceil to integer)
+			* step_gain_val = ""; (ceil to integer)
+			* default_gain = ""; (ceil to integer)
+			* Gain limits for mode
+			*
+			* exposure_factor = ""; (integer factor used for floating to fixed point conversion)
+			* min_exp_time = ""; (ceil to integer)
+			* max_exp_time = ""; (ceil to integer)
+			* step_exp_time = ""; (ceil to integer)
+			* default_exp_time = ""; (ceil to integer)
+			* Exposure Time limits for mode (us)
+			*
+			*
+			* min_hdr_ratio = "";
+			* max_hdr_ratio = "";
+			* HDR Ratio limits for mode
+			*
+			* framerate_factor = ""; (integer factor used for floating to fixed point conversion)
+			* min_framerate = "";
+			* max_framerate = "";
+			* step_framerate = ""; (ceil to integer)
+			* default_framerate = ""; (ceil to integer)
+			* Framerate limits for mode (fps)
+			*/
+			mode0 { // OV13B10_MODE_4208X3120
+				mclk_khz = "19200";
+				num_lanes = "4";
+				tegra_sinterface = "serial_e";
+				phy_mode = "DPHY";
+				discontinuous_clk = "yes";
+				dpcm_enable = "false";
+				cil_settletime = "0";
+
+				active_w = "4208";
+				active_h = "3120";
+				mode_type = "bayer";
+				pixel_phase = "rggb";
+				csi_pixel_bit_depth = "10";
+				readout_orientation = "0";
+				line_length = "4704";
+				inherent_gain = "1";
+				mclk_multiplier = "23.33";
+				pix_clk_hz = "448000000";
+
+				gain_factor = "16";
+				min_gain_val = "16"; /* 1.00x */
+				max_gain_val = "240"; /* 15.5x */
+				step_gain_val = "1";
+				default_gain = "16"; /* 1.00x */
+				min_hdr_ratio = "1";
+				max_hdr_ratio = "1";
+				framerate_factor = "1000000";
+				min_framerate = "2000000"; /* 2.0 fps */
+				max_framerate = "30000000"; /* 30.0 fps */
+				step_framerate = "1";
+				default_framerate = "30000000"; /* 30.0 fps */
+				exposure_factor = "1000000";
+				min_exp_time = "42"; /* us */
+				max_exp_time = "341676"; /* us */
+				step_exp_time = "1";
+				default_exp_time = "33251";/* us */
+				embedded_metadata_height = "0";
+			};
+
+			ports {
+				#address-cells = <1>;
+				#size-cells = <0>;
+
+				port@0 {
+					reg = <0>;
+					p2151_ov13b10_out2: endpoint {
+						port-index = <4>;
+						bus-width = <4>;
+						remote-endpoint = <&p2151_csi_in2>;
+					};
+				};
+			};
+		};
+	};
+
+	p2151_lens_ov7251@0 {
+		min_focus_distance = "0.041";
+		hyper_focal = "0.0";
+		focal_length = "0.83";
+		f_number = "2.0";
+		aperture = "0.0";
+	};
+
+	p2151_lens_ov13b10@1 {
+		min_focus_distance = "0.43";
+		hyper_focal = "0.0";
+		focal_length = "2.13";
+		f_number = "2.4";
+		aperture = "0.0";
+	};
+
+	tegra-camera-platform {
+		compatible = "nvidia, tegra-camera-platform";
+		/**
+		* Physical settings to calculate max ISO BW
+		*
+		* num_csi_lanes = <>;
+		* Total number of CSI lanes when all cameras are active
+		*
+		* max_lane_speed = <>;
+		* Max lane speed in Kbit/s
+		*
+		* min_bits_per_pixel = <>;
+		* Min bits per pixel
+		*
+		* vi_peak_byte_per_pixel = <>;
+		* Max byte per pixel for the VI ISO case
+		*
+		* vi_bw_margin_pct = <>;
+		* Vi bandwidth margin in percentage
+		*
+		* max_pixel_rate = <>;
+		* Max pixel rate in Kpixel/s for the ISP ISO case
+		*
+		* isp_peak_byte_per_pixel = <>;
+		* Max byte per pixel for the ISP ISO case
+		*
+		* isp_bw_margin_pct = <>;
+		* Isp bandwidth margin in percentage
+		*/
+		num_csi_lanes = <6>;
+		max_lane_speed = <1500000>;
+		min_bits_per_pixel = <10>;
+		vi_peak_byte_per_pixel = <2>;
+		vi_bw_margin_pct = <25>;
+		max_pixel_rate = <750000>;
+		isp_peak_byte_per_pixel = <5>;
+		isp_bw_margin_pct = <25>;
+
+		/**
+		* The general guideline for naming badge_info contains 3 parts, and is as follows,
+		* The first part is the camera_board_id for the module; if the module is in a FFD
+		* platform, then use the platform name for this part.
+		* The second part contains the position of the module, ex. “rear” or “front”.
+		* The third part contains the last 6 characters of a part number which is found
+		* in the module's specsheet from the vender.
+		*/
+		modules {
+			module0 {
+				badge = "ov7251_0_P2151X";
+				position = "0";
+				orientation = "1";
+				drivernode0 {
+					/* Declare PCL support driver (classically known as guid)  */
+					pcl_id = "v4l2_sensor";
+					/* Driver v4l2 device name */
+					devname = "ov7251 2-0061";
+					/* Declare the device-tree hierarchy to driver instance */
+					proc-device-tree = "/proc/device-tree/i2c@3180000/ov7251_a@61";
+				};
+				drivernode1 {
+					/* Declare PCL support driver (classically known as guid)  */
+					pcl_id = "v4l2_lens";
+					proc-device-tree = "/proc/device-tree/p2151_lens_ov7251@0/";
+				};
+			};
+			module1 {
+				badge = "ov7251_1_P2151X";
+				position = "1";
+				orientation = "1";
+				drivernode0 {
+					/* Declare PCL support driver (classically known as guid)  */
+					pcl_id = "v4l2_sensor";
+					/* Driver v4l2 device name */
+					devname = "ov7251 2-0062";
+					/* Declare the device-tree hierarchy to driver instance */
+					proc-device-tree = "/proc/device-tree/i2c@3180000/ov7251_b@62";
+				};
+				drivernode1 {
+					/* Declare PCL support driver (classically known as guid)  */
+					pcl_id = "v4l2_lens";
+					proc-device-tree = "/proc/device-tree/p2151_lens_ov7251@0/";
+				};
+			};
+			module2 {
+				badge = "ov13b10_2_P2151X";
+				position = "2";
+				orientation = "1";
+				drivernode0 {
+					/* Declare PCL support driver (classically known as guid)  */
+					pcl_id = "v4l2_sensor";
+					/* Driver v4l2 device name */
+					devname = "ov13b10 2-0036";
+					/* Declare the device-tree hierarchy to driver instance */
+					proc-device-tree = "/proc/device-tree/i2c@3180000/ov13b10_e@36";
+				};
+				drivernode1 {
+					/* Declare PCL support driver (classically known as guid)  */
+					pcl_id = "v4l2_lens";
+					proc-device-tree = "/proc/device-tree/p2151_lens_ov13b10@1/";
+				};
+			};
+		};
+	};
+};
diff --git a/kernel-dts/common/tegra194-fixed-regulator-p2151-0000.dtsi b/kernel-dts/common/tegra194-fixed-regulator-p2151-0000.dtsi
new file mode 100644
index 0000000..6b3b8c9
--- /dev/null
+++ b/kernel-dts/common/tegra194-fixed-regulator-p2151-0000.dtsi
@@ -0,0 +1,225 @@
+/*
+ * Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License, or
+ * (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ */
+
+#include <dt-bindings/gpio/tegra194-gpio.h>
+
+/ {
+	fixed-regulators {
+		p2515_vdd_3v3_cvb: regulator@101 {
+			compatible = "regulator-fixed";
+			reg = <101>;
+			regulator-name = "vdd-3v3-cvb";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+		};
+
+		p2515_vdd_1v8_cvb: regulator@102 {
+			compatible = "regulator-fixed";
+			reg = <102>;
+			regulator-name = "vdd-1v8-cvb";
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+		};
+
+		vdd_soc_fan: regulator@111 {
+			compatible = "regulator-fixed";
+			reg = <111>;
+			regulator-name = "vdd-soc-fan";
+			regulator-min-microvolt = <5000000>;
+			regulator-max-microvolt = <5000000>;
+		};
+
+		/* SOC_PWR_REQ->MOD_SLEEP->VDD_3V3_HDMI->VDD_5V0_HDMI_CN */
+		vdd_hdmi_5v0: regulator@112 {
+			compatible = "regulator-fixed";
+			reg = <112>;
+			regulator-name = "vdd-hdmi-5v0";
+			regulator-min-microvolt = <5000000>;
+			regulator-max-microvolt = <5000000>;
+			enable-active-high;
+		};
+
+		vdd_3v3_gps: reulator@21 {
+			compatible = "regulator-fixed";
+			reg = <21>;
+			regulator-name = "vdd-3v3-gps";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			gpio = <&gpio_expand2 16 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+
+		vdd_1v8_gps: reulator@22 {
+			compatible = "regulator-fixed";
+			reg = <22>;
+			regulator-name = "vdd-1v8-gps";
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			gpio = <&gpio_expand2 20 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+
+		// realtek wifi and bluetooth vbus
+		vdd_3v3_wlan_bt: reulator@35 {
+			compatible = "regulator-fixed";
+			reg = <35>;
+			regulator-name = "vdd-3v3-wlan-bt";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			gpio = <&gpio_expand2 21 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+
+		// INVALID, bcmdhd_wlan
+		vdd_1v8_wlan_en: reulator@36 {
+			compatible = "regulator-fixed";
+			reg = <36>;
+			regulator-name = "vdd-1v8-wlan-en";
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			gpio = <&gpio_expand1 11  GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+
+		// INVALID, bcmdhd_wlan
+		vdd_1v8_bt_en: reulator@37 {
+			compatible = "regulator-fixed";
+			reg = <37>;
+			regulator-name = "vdd-1v8-bt-en";
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			gpio = <&gpio_expand1 8 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+
+		rear_mcu_sensor: reulator@25 {
+			compatible = "regulator-fixed";
+			reg = <25>;
+			regulator-name = "rear-mcu-sensor";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			gpio = <&gpio_expand2 22 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+		head_mcu_sensor: reulator@26 {
+			compatible = "regulator-fixed";
+			reg = <26>;
+			regulator-name = "head-mcu-sensor";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			gpio = <&gpio_expand2 17 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+
+		bot_mcu_sensor: reulator@27 {
+			compatible = "regulator-fixed";
+			reg = <27>;
+			regulator-name = "bot-mcu-sensor";
+			regulator-min-microvolt = <3300000>;
+			regulator-max-microvolt = <3300000>;
+			gpio = <&gpio_expand1 14 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+
+		vbus_realsns_5v: reulator@28 {
+			compatible = "regulator-fixed";
+			reg = <28>;
+			regulator-name = "vbus-realsns-5v";
+			regulator-min-microvolt = <5000000>;
+			regulator-max-microvolt = <5000000>;
+			gpio = <&gpio_expand1 17 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+
+		vdd_case_fan: regulator@30 {
+			compatible = "regulator-fixed";
+			reg = <30>;
+			regulator-name = "vdd-case-fan";
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			gpio = <&gpio_expand2 3 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+
+		avdd_cam1_2v8: regulator@41 {
+			compatible = "regulator-fixed";
+			reg= <41>;
+			regulator-name = "avdd-cam1-2v8";
+			regulator-min-microvolt = <2800000>;
+			regulator-max-microvolt = <2800000>;
+			gpio = <&gpio_expand1 0 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+
+		avdd_cam2_2v8: regulator@42 {
+			compatible = "regulator-fixed";
+			reg= <42>;
+			regulator-name = "avdd-cam2-2v8";
+			regulator-min-microvolt = <2800000>;
+			regulator-max-microvolt = <2800000>;
+			gpio = <&gpio_expand1 1 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+
+		avdd_cam3_2v8: regulator@43 {
+			compatible = "regulator-fixed";
+			reg= <43>;
+			regulator-name = "avdd-cam3-2v8";
+			regulator-min-microvolt = <2800000>;
+			regulator-max-microvolt = <2800000>;
+			gpio = <&gpio_expand1 2 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+
+		iovdd_cam_1v8: regulator@44{
+			compatible = "regulator-fixed";
+			reg= <44>;
+			regulator-name = "iovdd-cam-1v8";
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			gpio = <&gpio_expand1 3 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+
+		dvdd_codec_amp_1v8: regulator@50{
+			compatible = "regulator-fixed";
+			reg= <50>;
+			regulator-name = "dvdd-codec-amp-1v8";
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			gpio = <&gpio_expand2 18 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+
+		avdd_codec_1v8: regulator@51{
+			compatible = "regulator-fixed";
+			reg= <51>;
+			regulator-name = "avdd-codec-1v8";
+			regulator-min-microvolt = <1800000>;
+			regulator-max-microvolt = <1800000>;
+			gpio = <&gpio_expand2 19 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+
+		pvdd_amp_12v: regulator@52{
+			compatible = "regulator-fixed";
+			reg= <52>;
+			regulator-name = "pvdd-amp-12v";
+			regulator-min-microvolt = <12000000>;
+			regulator-max-microvolt = <12000000>;
+			gpio = <&gpio_expand1 7 GPIO_ACTIVE_HIGH>;
+			enable-active-high;
+		};
+	};
+};
diff --git a/kernel-dts/common/tegra194-p2151-0000.dtsi b/kernel-dts/common/tegra194-p2151-0000.dtsi
new file mode 100644
index 0000000..ea8df64
--- /dev/null
+++ b/kernel-dts/common/tegra194-p2151-0000.dtsi
@@ -0,0 +1,456 @@
+/*
+ * Copyright (c) 2019-2021, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ */
+#include <dt-bindings/gpio/tegra194-gpio.h>
+#include "dt-bindings/input/input.h"
+#include "tegra194-fixed-regulator-p2151-0000.dtsi"
+#include "tegra194-p2151-disp.dtsi"
+#include "tegra194-audio-p3668.dtsi"
+#include "tegra194-camera-p2151.dtsi"
+
+/ {
+	gpio-keys {
+		compatible = "gpio-keys";
+		gpio-keys,name = "gpio-keys";
+
+		forcerecovery {
+			label = "force-recovery";
+			gpios = <&tegra_main_gpio TEGRA194_MAIN_GPIO(G, 0) GPIO_ACTIVE_LOW>;
+			linux,code = <BTN_1>;
+		};
+
+		power_key {
+			label = "power-key";
+			gpios = <&tegra_aon_gpio TEGRA194_AON_GPIO(EE, 4) GPIO_ACTIVE_LOW>;
+			linux,code = <KEY_POWER>;
+			gpio-key,wakeup;
+		};
+	};
+
+#if TEGRA_PWM_FAN_DT_VERSION == DT_VERSION_2
+	pwm-fan {
+		compatible = "pwm-fan";
+		status = "okay";
+		#pwm-cells = <1>;
+		pwms = <&tegra_pwm8 0 45334>;
+		fan-supply = <&vdd_soc_fan>;
+		profiles {
+			default = "quiet";
+			quiet {
+				cooling-levels = <255 178 135 95 0 0 0 0 0 0>;
+			};
+			cool {
+				cooling-levels = <255 178 135 95 0 0 0 0 0 0>;
+			};
+		};
+	};
+	pwm-fan-case {
+		compatible = "pwm-fan";
+		status = "okay";
+		#pwm-cells = <1>;
+		pwms = <&tegra_pwm6 0 45334>;
+		fan-supply = <&vdd_case_fan>;
+		profiles {
+			default = "quiet";
+			quiet {
+				cooling-levels = <255 178 135 95 0 0 0 0 0 0>;
+			};
+			cool {
+				cooling-levels = <255 178 135 95 0 0 0 0 0 0>;
+			};
+		};
+	};
+#else
+	// soc fan
+	pfsd {
+		pwm_id = <8>;
+		pwm_gpio = <&tegra_main_gpio TEGRA194_MAIN_GPIO(R, 0) GPIO_ACTIVE_LOW>;
+	};
+	pwm-fan {
+		pwms = <&tegra_pwm8 0 45334>;
+		vdd-fan-supply = <&p3509_vdd_fan>;
+		profiles {
+			default = "quiet";
+			quiet {
+				state_cap = <4>;
+				active_pwm = <0 130 160 200 255 255 255 255 255 255>;
+			};
+			cool {
+				state_cap = <4>;
+				active_pwm = <0 140 170 200 255 255 255 255 255 255>;
+			};
+		};
+	};
+	// case fan
+	pwm_fan_case_shared_data: pfsd-case {
+		num_resources = <0>;
+		secret = <47>;
+		active_steps = <10>;
+		active_rpm = <0 1000 2000 3000 4000 5000 6000 7000 10000 11000>;
+		rpm_diff_tolerance = <2>;
+		active_rru = <40 2 1 1 1 1 1 1 1 1>;
+		active_rrd = <40 2 1 1 1 1 1 1 1 1>;
+		state_cap_lookup = <2 2 2 2 3 3 3 4 4 4>;
+		pwm_period = <45334>;
+		pwm_id = <6>;
+		pwm_polarity = <PWM_POLARITY_INVERTED>;
+		suspend_state = <1>;
+		step_time = <100>; /* mesecs */
+		state_cap = <7>;
+		active_pwm_max = <256>;
+		tach_period = <1000>;
+		pwm_gpio = <&tegra_aon_gpio TEGRA194_MAIN_GPIO(H, 1) GPIO_ACTIVE_LOW>;
+		tach_gpio = <&tegra_main_gpio TEGRA194_MAIN_GPIO(Q, 2) GPIO_ACTIVE_LOW>;
+	};
+	pwm-fan-case {
+		compatible = "pwm-fan";
+		status = "okay";
+		#pwm-cells = <1>;
+		pwms = <&tegra_pwm6 0 45334>;
+		vdd-fan-supply = <&p3509_vdd_fan>;
+		shared_data = <&pwm_fan_case_shared_data>;
+		profiles {
+			default = "quiet";
+			quiet {
+				state_cap = <4>;
+				active_pwm = <0 77 120 160 255 255 255 255 255 255>;
+			};
+			cool {
+				state_cap = <4>;
+				active_pwm = <0 77 120 160 255 255 255 255 255 255>;
+
+			};
+		};
+	};
+#endif
+	thermal-fan-est {
+		profiles {
+			default = "quiet";
+			quiet {
+				active_trip_temps = <0 46000 60000 68000 76000
+						140000 150000 160000 170000 180000>;
+				active_hysteresis = <0 8000 8000 7000 7000
+						0 0 0 0 0>;
+			};
+			cool {
+				active_trip_temps = <0 35000 45000 53000 61000
+						140000 150000 160000 170000 180000>;
+				active_hysteresis = <0 8000 8000 7000 7000
+						0 0 0 0 0>;
+			};
+		};
+	};
+	realsense_switch {
+		compatible = "reg-userspace-consumer";
+		regulator-name = "realsense_switch";
+		regulator-supplies = "default";
+		default-supply = <&vbus_realsns_5v>;
+		status = "okay";
+	};
+	wifi_bt_switch {
+		compatible = "reg-userspace-consumer";
+		regulator-boot-on;
+		regulator-name = "wifi_bt_switch";
+		regulator-supplies = "default";
+		default-supply = <&vdd_3v3_wlan_bt>;
+		status = "okay";
+	};
+	gps_switch {
+		compatible = "reg-userspace-consumer";
+		regulator-name = "gps_switch";
+		regulator-supplies = "vdd-3v3", "vdd-1v8";
+		vdd-3v3-supply = <&vdd_3v3_gps>;
+		vdd-1v8-supply = <&vdd_1v8_gps>;
+		status = "okay";
+	};
+	mcu_sensor_switch {
+		compatible = "reg-userspace-consumer";
+		regulator-name = "mcu_sensor_switch";
+		regulator-boot-on;
+		regulator-supplies = "head", "bot", "rear";
+		head-supply = <&head_mcu_sensor>;
+		bot-supply = <&bot_mcu_sensor>;
+		rear-supply = <&rear_mcu_sensor>;
+		status = "okay";
+	};
+	serial@3110000 {
+		status = "okay";
+	};
+	// disable THS4, AGX for M2 only
+	serial@3140000 {
+		status = "disabled";
+	};
+	// Case Tech
+	tachometer@39c0000 {
+		status = "okay";
+	};
+	// Case Fan
+	tegra_pwm6: pwm@32d0000 {
+		status = "okay";
+	};
+	// SoC Fan
+	tegra_pwm8: pwm@32f0000 {
+		status = "okay";
+	};
+
+	sdhci@3400000 {
+		status = "disabled";
+	};
+
+	gpio@2200000 {
+		/* gpio-name given as COL(10) x ROW(24),
+		   Tegra-GPIO starts from 288 */
+		gpio-line-names =
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "CASE_FAN_PWM", "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "",             "",             "",             "GPIO13",       "",             "",             "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "GPIO_EXP1_INT",
+		"CASE_FAN_TACH","TOUCH_INT",    "",             "GPIO01",       "GPIO11",       "",             "SOC_FAN_PWM",  "",             "",             "",
+		"UART1_RTS",    "UART1_CTS",    "",             "",             "",             "",             "",             "",             "GPIO09",       "",
+		"",             "",             "",             "",             "",             "",             "",             "I2S0_SCLK",    "I2S0_DOUT",    "I2S0_DIN",
+		"I2S0_FS",      "",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "SPI1_SCK",     "SPI1_MISO",    "SPI1_MOSI",    "SPI1_CS0",     "SPI1_CS1",     "",             "",             "",
+		"",             "USB0_VBUS_DET","",             "SPI0_SCK",     "SPI0_MISO",    "SPI0_MOSI",    "SPI0_CS0",     "SPI0_CS1",     "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "",             "";
+	};
+
+	gpio@c2f0000 {
+		/* gpio-name given as COL(10) x ROW(2),
+		   Tegra-AON-GPIO starts from 248 */
+		gpio-line-names =
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "GPIO_EXP12_RST",
+		"GPIO_EXP2_INT","",             "",             "",             "",             "",             "",             "",             "",             "",
+		"",             "",             "",             "",             "",             "",             "",             "",             "",             "";
+	};
+
+	// NX     Ball  CVM/NX       NX/P3509           P2151
+	// GPIO00 PZ.01 USB_VBUS_EN0                    [USB0_VBUS_DET]
+	// GPIO01 PQ.05 EXTPERIPH3                      [OV7251_L_MCLK]
+	// GPIO02 PQ.03              (BT_M2_WAKE_AP)    [TOUCH_INT]
+	// GPIO03 CC.00              (BT_M2_EN)
+	// GPIO04 CC.01              (PWR_LED_CTRL)
+	// GPIO05 CC.02              (W_DISKABLE1_CTRL)
+	// GPIO06 CC.03              (CAM_MUX_SEL)      [GPIO_EXP12_RST]
+	// GPIO07 PR.00 PWM8                            [SOC_FAN_PWM]
+	// GPIO08 PQ.02 SDMMC_CD                        [CASE_FAN_TACH]
+	// GPIO09 PS.04 AUD_MCLK
+	// GPIO10 PQ.01              (M2E_ALERT)        [GPIO_EXP1_INT]
+	// GPIO11 PQ.06 EXTPERIPH4
+	// GPIO12 CC.04 PWM4                            [GPIO_EXP2_INT]
+	// GPIO13 PN.01 PWM1
+	// GPIO14 PH.01 PWM6                            [CASE_FAN_PWM]
+
+	i2c@3160000 {
+		status = "okay";
+
+		gpio_expand2: tca6424@22 {
+			compatible = "ti,tca6424"; // gpio-pca953x
+			reg = <0x22>;
+			// share a reset pin for two expanders is a dangerous design
+			// reset-gpio = <&tegra_aon_gpio TEGRA194_AON_GPIO(CC, 3) GPIO_ACTIVE_HIGH>;
+			gpio-controller;
+			#gpio-cells = <2>;
+			status = "okay";
+			interrupt-parent = <&tegra_aon_gpio>;
+			interrupts = <TEGRA194_AON_GPIO(CC, 4) IRQ_TYPE_LEVEL_LOW>;
+			interrupt-controller;
+			#interrupt-cells = <2>;
+			// x_LABEL is NOT sure
+			gpio-line-names =
+				"x_PROG_UTRA_F",   "x_NULL",             "x_RSTN_UTRA_F",      "CASE_FAN_EN",
+				"x_RSTN_HUB",      "x_RSTN_UTRA_R",      "x_INT_TOUCHPAD",     "TOUCHPAD_NRST",
+				"x_LED_DRIVER_EN", "x_NULL",             "GPS_NSTANDBY",       "x_H_HOST_REQ",
+				"x_MCU_RDY_GPS",   "x_MCU_REQ_GPS",      "FLASH_LED",          "TORCH_LED",
+				"GPS_3V3_EN",      "HEAD_SENSOR_3V3_EN", "SND_PA_1V8_EN",      "SND_CODEC_1V8_EN",
+				"GPS_1V8_EN",      "WLAN_BT_3V3_EN",     "REAR_SENSOR_3V3_EN", "x_SENSOR_1V8_EN";
+			// 03 CASE_FAN_EN: PROG_UTRA_R, unknown
+			// 07 TOUCHPAD_NRST: Synaptics touchpad reset, Low on reset
+			// 10 GPS_NSTANDBY: AP6155G/BCM4775 nstandby, Low on standby. Exposed to user space
+			// 14 FLASH_LED: Flash LED, high enable
+			// 15 TORCH_LED: Torch LED, high enable
+			// 16 GPS_3V3_EN: AP6155G/BCM4775 3v3 power, high enable
+			// 17 HEAD_SENSOR_3V3_EN: Head sensor 3v3 poewr, high enable
+			// 18 SND_PA_1V8_EN: TAS5805 1v8 power, high enable
+			// 19 SND_CODEC_1V8_EN: RT5680 1v8 power, high enable
+			// 20 GPS_1V8_EN: AP6155G/BCM4775 1v8 power, high enable
+			// 21 WLAN_BT_3V3_EN: 6221B/rtl8821cu 3v3 power, high enable
+			// 22 REAR_SENSOR_3V3_EN: Rear sensor 3v3 poewr, high enable
+		};
+
+		gpio_expand1: tca6424@23 {
+			compatible = "ti,tca6424";
+			reg = <0x23>;
+			// reset-gpio = <&tegra_aon_gpio TEGRA194_AON_GPIO(CC, 3) GPIO_ACTIVE_HIGH>;
+			gpio-controller;
+			#gpio-cells = <2>;
+			interrupt-parent = <&tegra_main_gpio>;
+			interrupts = <TEGRA194_MAIN_GPIO(Q, 1) IRQ_TYPE_LEVEL_LOW>;
+			interrupt-controller;
+			#interrupt-cells = <2>;
+			status = "okay";
+			// x_LABEL is NOT sure
+			gpio-line-names =
+				"CAM1_2V8_EN",      "CAM2_2V8_EN",         "CAM3_2V8_EN",       "CAMERA_1V8_EN",
+				"x_SEL1_MODE",      "x_RESET_CODEC",       "x_LDO1_EN_CODEC",   "SND_PA_PDN",
+				"BT_1V8_EN",        "x_HOST_WAKE_BT",      "x_BT_WAKE_HOST",    "WLAN_1V8_EN",
+				"x_WLAN_WAKE_HOST", "x_BOT_SENSOR_2V8_EN", "BOT_SENSOR_3V3_EN", "x_INT_TUSB322",
+				"x_VBUS_MOTION_EN", "REALSENSE_VBUS_EN",   "x_NULL",            "x_CAM3_STROBE",
+				"CAM3_PWDN",        "x_TOF_EN",            "x_CAM2_STROBE",     "x_TOF_GPIO1";
+			// 00 CAM1_2V8_EN: Canera1 2v8 power, high enable
+			// 01 CAM2_2V8_EN: Canera2 2v8 power, high enable
+			// 02 CAM3_2V8_EN: Canera3 2v8 power, high enable
+			// 03 CAMERA_1V8_EN: Canera1/2/3 1v8 power, high enable
+			// 07 SND_PA_PDN: TAS5805 power down, low power down, high enable
+			// 08 BT_EN: 6221B/rtl8821cu bluetooth enable, high enable
+			// 11 WLAN_EN: 6221B/rtl8821cu wifi enable, high enable
+			// 14 BOT_SENSOR_3V3_EN: Botton sensor(optical flow) 3v3 power, high enable
+			// 17 REALSENSE_VBUS_EN: Realsense vbus power, high enable
+			// 20 CAM3_PWDN: Canera3 power down, high enable
+		};
+
+
+		tas5805m: tas5805m@2c {
+			compatible = "ti,tas5805m";
+			reg = <0x2c>;
+			dvdd-supply = <&dvdd_codec_amp_1v8>;
+			pvdd-supply = <&pvdd_amp_12v>;
+			status = "disabled";
+		};
+
+		rt5680: rt5680@2d {
+			compatible = "realtek,rt5679";
+			reg = <0x2d>;
+			dvdd-supply = <&dvdd_codec_amp_1v8>;
+			avdd-supply = <&avdd_codec_1v8>;
+			status = "disabled";
+		};
+	};
+
+#define	CAM1_PWDN	TEGRA194_MAIN_GPIO(P, 4)
+#define	CAM2_PWDN	TEGRA194_MAIN_GPIO(P, 5)
+#define	CAM3_PWDN	20
+
+	i2c@3180000 {
+		ov7251_a@61 { // OV7251_R
+			clocks = <&bpmp_clks TEGRA194_CLK_EXTPERIPH2>,
+				 <&bpmp_clks TEGRA194_CLK_PLLP_OUT0>;
+			clock-names = "extperiph2", "pllp_grtba";
+			mclk = "extperiph2";
+			clock-frequency = <24000000>;
+			pwdn-gpios = <&tegra_main_gpio CAM2_PWDN GPIO_ACTIVE_HIGH>;
+			vana-supply = <&avdd_cam2_2v8>;
+			vif-supply = <&iovdd_cam_1v8>;
+			status = "okay";
+		};
+		ov7251_b@62 { // OV7251_L
+			clocks = <&bpmp_clks TEGRA194_CLK_EXTPERIPH3>,
+				 <&bpmp_clks TEGRA194_CLK_PLLP_OUT0>;
+			clock-names = "extperiph3", "pllp_grtba";
+			mclk = "extperiph3";
+			clock-frequency = <24000000>;
+			pwdn-gpios = <&gpio_expand1 CAM3_PWDN GPIO_ACTIVE_HIGH>;
+			vana-supply = <&avdd_cam3_2v8>;
+			vif-supply = <&iovdd_cam_1v8>;
+			status = "okay";
+		};
+		ov13b10_e@36 {
+			clocks = <&bpmp_clks TEGRA194_CLK_EXTPERIPH1>,
+				 <&bpmp_clks TEGRA194_CLK_PLLP_OUT0>;
+			clock-names = "extperiph1", "pllp_grtba";
+			mclk = "extperiph1";
+			clock-frequency = <19200000>;
+			pwdn-gpios = <&tegra_main_gpio CAM1_PWDN GPIO_ACTIVE_HIGH>;
+			vana-supply = <&avdd_cam1_2v8>;
+			vif-supply = <&iovdd_cam_1v8>;
+			status = "okay";
+		};
+	};
+
+	i2c@31e0000 {
+		status = "okay";
+
+		synaptics_dsx@20 {
+			compatible = "synaptics,dsx-i2c";
+			reg = <0x20>;
+			status = "disabled";
+			interrupt-parent = <&tegra_main_gpio>;
+			interrupts = <TEGRA194_MAIN_GPIO(Q, 3)  0x2002>;
+			synaptics,ub-i2c-addr = <0x2c>;
+			synaptics,irq-gpio = <&tegra_main_gpio TEGRA194_MAIN_GPIO(Q, 3) 0x2002>; // IRQF_ONESHOT|IRQF_TRIGGER_FALLING
+			synaptics,reset-gpio = <&gpio_expand2 7 0x00>;
+			synaptics,irq-on-state = <0>;
+			synaptics,power-delay-ms = <200>;
+			synaptics,reset-delay-ms = <200>;
+			synaptics,reset-on-state = <0>;
+			synaptics,reset-active-ms = <20>;
+			// uncomment and update line below as appropriate if there are 0D buttons
+			// synaptics,cap-button-codes = <102 158>;// use in debug, should be comment in final release!!!!
+			// uncomment and update lines below as appropriate if there are virtual buttons
+			// synaptics,vir-button-codes = <102 100 900 100 60 158 300 900 100 60>;
+			// synaptics,max-y-for-2d = <800>;
+	};
+	};
+
+	bcm4775 {
+		compatible = "bcm4775";
+		vdd-supply = <&vdd_1v8_gps>;
+		vddgps-supply = <&vdd_3v3_gps>;
+		nstandby-gpio = <&gpio_expand2 10 GPIO_ACTIVE_LOW>;
+		status = "disabled";
+	};
+
+	leds {
+		flash {
+			gpios = <&gpio_expand2 14 GPIO_ACTIVE_HIGH>;
+			default-state = "off";
+		};
+
+		torch {
+			gpios = <&gpio_expand2 15 GPIO_ACTIVE_HIGH>;
+			default-state = "off";
+		};
+	};
+
+	ethernet@2490000 {
+		nvidia,skip_mac_reset = "0";
+	};
+
+	// PCIe C4
+	pcie@14160000 {
+		status = "disabled";
+	};
+
+	aconnect@2a41000 {
+		status = "disabled";
+	};
+
+	spdif_dit {
+		status = "disabled";
+	};
+
+	tegra_sound: sound {
+		status = "disabled";
+	};
+};
diff --git a/kernel-dts/common/tegra194-p2151-disp.dtsi b/kernel-dts/common/tegra194-p2151-disp.dtsi
new file mode 100644
index 0000000..d5a0c8d
--- /dev/null
+++ b/kernel-dts/common/tegra194-p2151-disp.dtsi
@@ -0,0 +1,61 @@
+/*
+ * tegra194-p2151-disp.dtsi:
+ *
+ * Copyright (c) 2019, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ *
+ *
+ * athena uses 1 display heads out of the 4 available on t194:
+ *	fb0: Head0->SOR1->HDMI
+ * Each display head is assigned three windows each.
+ */
+
+#include <dt-bindings/display/tegra-dc.h>
+#include <dt-bindings/display/tegra-panel.h>
+#include <t19x-common-platforms/tegra194-hdmi.dtsi>
+#include <t19x-common-platforms/tegra194-dp.dtsi>
+#include "tegra194-fixed-regulator-p2151-0000.dtsi"
+#include "tegra194-spmic-p3668.dtsi"
+
+&head0 {
+	status = "okay";
+	nvidia,fb-bpp = <32>;
+	nvidia,fbmem-size = <16588800>; /* 1080P (1920*1080) 32bpp double buffered */
+	nvidia,fb-flags = <TEGRA_FB_FLIP_ON_PROBE>;
+	win-mask = <0x7>;
+	nvidia,fb-win = <0>;
+	nvidia,dc-connector = <&sor1>;
+	nvidia,dc-flags = <TEGRA_DC_FLAG_ENABLED>;
+	avdd_hdmi-supply = <&p3668_spmic_sd0>; /* 1v0 */
+	avdd_hdmi_pll-supply = <&p3668_spmic_sd1>; /* 1v8 */
+	vdd_hdmi_5v0-supply = <&vdd_hdmi_5v0>; /* 5v0 */
+};
+
+&sor1 {
+	status = "okay";
+	nvidia,active-panel = <&sor1_hdmi_display>;
+	nvidia,sor-hdcp-not-supported;
+};
+
+&sor1_hdmi_display {
+	status = "okay";
+	disp-default-out {
+		nvidia,out-flags = <TEGRA_DC_OUT_HOTPLUG_LOW>;
+	};
+};
+
+&dpaux1 {
+	status = "okay";
+};
+
+&tegra_cec {
+	status = "okay";
+};
diff --git a/kernel-dts/tegra194-p3668-0001-p2151-0000.dts b/kernel-dts/tegra194-p3668-0001-p2151-0000.dts
new file mode 100644
index 0000000..eb7e0a6
--- /dev/null
+++ b/kernel-dts/tegra194-p3668-0001-p2151-0000.dts
@@ -0,0 +1,26 @@
+/*
+ * Top level DTS file for CVM:P3668-0001 and CVB:P3509-0000.
+ *
+ * Copyright (c) 2019-2020, NVIDIA CORPORATION.  All rights reserved.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; version 2 of the License.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License for
+ * more details.
+ */
+
+/dts-v1/;
+#include "common/tegra194-p3668-common.dtsi"
+#include "common/tegra194-p2151-0000.dtsi"
+
+/ {
+	nvidia,dtsfilename = "p3668-0001-p2151-0000.dts";
+	nvidia,dtbbuildtime = __DATE__, __TIME__;
+
+	model = "Xiaomi Cyberdog";
+	compatible = "nvidia,p2151-0000+p3668-0001", "nvidia,tegra194";
+};
-- 
2.50.1 (Apple Git-155)

